### ISSUE TITLE
Rule proposal: no-unnecessary-optional-prop — detect optional props that are always provided

### ISSUE THREAD

**Issue by reteps** [Rel1cx/eslint-react#1598]
## What this rule would do

Flag React component props that are marked as optional (`?`) but are provided at every JSX call site in the project. This suggests the prop should be made required.

## Example

```tsx
// Component definition
function Button({ label, onClick }: { label: string; onClick?: () => void }) {
  //                                                   ^ optional but always provided
}

// Every usage provides onClick:
<Button label="Save" onClick={handleSave} />
<Button label="Cancel" onClick={handleCancel} />
<Button label="Delete" onClick={handleDelete} />
```

After fixing:

```tsx
function Button({ label, onClick }: { label: string; onClick: () => void }) {
  // ...
}
```

## Motivation

Unnecessarily optional props weaken type safety — the component body must handle `undefined` for a value that is never actually omitted. This leads to unnecessary nullish checks, default values that are never used, and a misleading API surface. Over time as codebases evolve, props that were once optional become effectively required, but the type definition isn't updated.

This is similar in spirit to existing rules like `no-unused-props` (defined but never used) — this would catch the inverse-adjacent case (optional but always provided).

## Considerations

- The rule would need whole-program analysis across files, similar to `no-unused-props`.
- It should only flag props where **every** JSX usage provides the prop. Spread props (`{...props}`) at a call site could be treated as ambiguous and cause the rule to skip that component.
- Components that are exported from the project's public API boundary could be excluded, since external consumers may rely on the prop being optional.
- Could support an option for a minimum call-site threshold (e.g., only flag if the component is used at least N times, to avoid false positives on rarely-used components).

**Comment by reteps** [comment]
Closing to re-file using the rule request template.
